File: Binary Tree Preorder Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        vector<int> res;
        if (root == NULL) return res;
        
        stack<TreeNode *> st;
        while (root) {
            st.push(root);
            res.push_back(root->val);
            root = root->left;
        }
        
        while (!st.empty()) {
            root = st.top()->right;
            st.pop();
            while (root) {
                st.push(root);
                res.push_back(root->val);
                root = root->left;
            }
        }
        
        return res;
    }
};

File: Binary Tree Preorder Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> PreorderTraversal(TreeNode root) {
        var sequence = new List<int>();
        if (root == null) {
            return sequence;
        }
        
        var stack = new Stack<TreeNode>();

        
        while (true) {
            while (root != null) {
                stack.Push(root);
                sequence.Add(root.val);
                root = root.left;
            }
            
            if (stack.Count == 0) {
                break;
            }
            
            root = stack.Pop();
            root = root.right;
        }
        
        return sequence;
    }
}
